userspace/async/async_key_value_source.h

```diff
@@ -0,0 +1,280 @@
+	 */
+	bool is_running() const;
+
+protected:
+	/**
+	 * Stops the thread assocaited with this async_key_value_source, if
+	 * it is running; otherwise, does nothing.  The only use for this is
+	 * in a destructor to ensure that the async thread stops when the
+	 * object is destroyed.
+	 */
+	void stop();
+
+	/**
+	 * Dequeues an entry from the request queue and returns it in the given
+	 * key.  Concrete subclasses will call this method to get the next key
+	 * for which to collect values.
+	 *
+	 * @returns true if there was a key to dequeue, false otherwise.
+	 */
+	bool dequeue_next_key(key_type& key);
+
+	/**
+	 * Get the (potentially partial) value for the given key.
+	 *
+	 * @param[in] key The key whose value is needed.
+	 *
+	 * @returns the value associated with the given key.
+	 */
+	value_type get_value(const key_type& key);
+
+	/**
+	 * Stores a value for the given key.  Concrete subclasses will call
+	 * this method from their run_impl() method to save (or otherwise
+	 * notifiy the client about) an available value.
+	 *
+	 * @param[in] key   The key for which the client asked for the value.
+	 * @param[in] value The collected value.
+	 */
+	void store_value(const key_type& key, const value_type& value);
+
+	/**
+	 * Concrete subclasses must override this method to perform the
+	 * asynchronous value lookup.  The implementation should:
+	 *
+	 * <ul>
+	 * <li>Loop while dequeue_next_key() is true.</li>
+	 * <li>Get any existing value for that key using get_value()</li>
+	 * <li>Do whatever work is necessary to lookup the value associated
+	 *     with that key.</li>
+	 * <li>Call store_value to store the updated value, and to
+	 *     notify any client code waiting on that data.</li>
+	 * </ul>
+	 */
+	virtual void run_impl() = 0;
+
+private:
+	/**
+	 * Holds information associated with a single lookup() request.
+	 */
+	struct lookup_request
+	{
+		lookup_request():
+			m_available(false),
+			m_value(),
+			m_available_condition(),
+			m_callback(),
+			m_start_time(std::chrono::steady_clock::now())
+		{ }
+
+		/** Is the value here available? */
+		bool m_available;
+
+		/** The value for a key. */
+		value_type m_value;
+
+		/** Block in lookup() waiting for a sync response. */
+		std::condition_variable m_available_condition;
+
+		/**
+		 * A optional client-specified callback handler for async
+		 * response notification.
+		 */
+		callback_handler m_callback;
+
+		/** The time at which this request was made. */
+		std::chrono::time_point<std::chrono::steady_clock> m_start_time;
+	};
+
+	typedef std::map<key_type, lookup_request> value_map;
+
+	/**
+	 * The entry point of the async thread, which blocks waiting for work
+	 * and dispatches work to run_impl().
+	 */
+	void run();
+
+	/**
+	 * Remove any entries that are older than the time-to-live.
+	 */
+	void prune_stale_requests();
+
+	uint64_t m_max_wait_ms;
+	uint64_t m_ttl_ms;
+	std::thread m_thread;
+	bool m_running;
+	bool m_terminate;
+	mutable std::mutex m_mutex;
```

Comment:
as we talked about, unfortuantely this is in the header, but I'd like some commends on what this mutex and the subsequent CVs are protecting